src/main/java/org/jpeek/calculus/java/Ccm.java

```diff
@@ -52,57 +61,163 @@ public XML node(
+                    );
+                    tag.appendChild(sub);
+                }
+                meta.appendChild(tag);
             }
+            final Node repr = skeleton.node();
+            final Node text = repr.getFirstChild().getOwnerDocument()
+                .importNode(doc.createTextNode("\n"), true);
+            final Node node = repr.getFirstChild().getOwnerDocument()
+                .importNode(meta, true);
+            repr.getFirstChild().appendChild(text);
+            repr.getFirstChild().appendChild(node);
+            return new XMLDocument(repr);
+        } catch (final ParserConfigurationException ex) {
+            throw new IllegalStateException(ex);
         }
-        return transformed;
     }
 
     /**
-     * Updates the xml node of the class with proper NCC value.
-     * @param skeleton XML Skeleton
-     * @param pack Package name
-     * @param clazz Class node in the resulting xml
-     * @todo #449:30min Implement NCC calculation with `XmlGraph` and use this
-     *  class to fix CCM metric (see issue #449). To do this, this class, once
-     *  it works correctly, should be integrated with XSL based calculuses in
-     *  `XslReport` (see `todo #449` in Calculus). Write a test to make sure
-     *  the metric is calculated correctly. Also, decide whether the
-     *  whole CCM metric should be implemented in Java, or only the NCC part.
-     *  Update this `todo` accordingly.
+     * Adds NCC (Number of Component Connections) tag to the XML document.
+     * This method calculates the NCC for a given class and adds it as a tag to the XML document.
+     * @param doc The XML document to which the NCC tag will be added.
+     * @param clazz The XML representation of the class.
+     * @param params Parameters for the computation (unused).
+     * @return The NCC node.
      */
-    private static void updateNcc(
-        final XML skeleton, final String pack, final XML clazz
+    private static Node addNccTag(final Document doc, final XML clazz,
+        final Map<String, Object> params
     ) {
-        throw new UnsupportedOperationException(
-            new Joined(
-                "",
-                skeleton.toString(),
-                pack,
-                clazz.toString()
-            ).toString()
-        );
+        final Element ncc = doc.createElement("ncc");
+        ncc.appendChild(doc.createTextNode(calculateComponents(clazz, params).toString()));
+        return ncc;
+    }
+
+    /**
+     * Calculates the number of components for a given class.
+     * This method calculates the number of components for a class using the Union-Find algorithm.
+     * @param clazz The XML representation of the class.
+     * @param params Parameters for the computation.
+     * @return The number of components.
+     */
+    private static Integer calculateComponents(final XML clazz, final Map<String, Object> params) {
```

Comment:
May be it is better to move all logic that operates with graph to separate class and do not create so many static methods.